### PR TITLE
Fixes #1351 - Add functionality to project selector to create new project

### LINKED ISSUE #1351: Add functionality to project selector to create new project
~~The project list should contain a special item on the top to create a new GCP project. It will open the Cloud Console in the external browser so that the user can create the project.~~

### PR DESCRIPTION
**Pull request by akerekes**
Adds a simple dialog that displays a link which opens a browser to Cloud Console. On the opened page the user can create a new project with an App Engine application. After the user closes the create project dialog, we reload the project list assuming that the new project has been created and the user wants to select it.

Known issues:
 - the page opened is not only for creating a project but also adding App Engine application to an existing project, so the user must know that she needs to click on the `Create Project` button
 - if the user belongs to an organization and clicks on the `Create Project` button, she may get an error page that she's not allowed to create a project outside of an organization. In this case, she needs to find how to create a project in the console (using the project selector at the top of the page). The project created this way will not have an App Engine application, but this will be recognized by the project selector, however this flow may be frustrating to the user.

I don't expect this dialog to remain this simple for long, we should let the users create a project directly from the IDE ASAP.

image

image

### CODE REVIEW

**Comment by akerekes** [comment]
@elharo I'm expecting this dialog to be only temporary, and replace its content soon with widgets to create the project in the IDE, so in that sense this is a placeholder until then.

File: plugins/com.google.cloud.tools.eclipse.projectselector/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -40,6 +40,7 @@ Import-Package: com.google.api.client.auth.oauth2,
  org.eclipse.jface.dialogs,
  org.eclipse.jface.layout,
  org.eclipse.jface.viewers,
+ org.eclipse.jface.window,
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.projectselector/META-INF/MANIFEST.MF]: I don't think this is added.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties [hunk 2]
```diff
@@ -40,6 +40,11 @@ tooltip.stop.previous.version=If checked, stops the previously running version w
  new version that receives all traffic.
 tooltip.staging.bucket=The Google Cloud Storage bucket used to stage files for deployment. \
  If not specified, the application''s default code bucket is used.
+
+#Project selector
+projectselector.create.newproject=Create GCP project...
+projectselector.dialog.createproject.title=Create new GCP project
+projectselector.dialog.createproject.message=You can create a new GCP project in the <a href="{0}">Cloud Console</a>
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties]: Let's put a period at the end.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java [hunk 3]
```diff
@@ -284,15 +286,37 @@ private void createProjectIdSection() {
...
+    GridDataFactory.fillDefaults().grab(true, false).hint(SWT.DEFAULT, 200)
+        .applyTo(projectSelector);
+
+    Button createProjectButton = new Button(projectSelectorComposite, SWT.NONE);
+    createProjectButton.setText(Messages.getString("projectselector.create.newproject"));
+    GridDataFactory.swtDefaults().align(SWT.FILL, SWT.BEGINNING).applyTo(createProjectButton);
+    createProjectButton.addSelectionListener(new SelectionListener() {
+
+      @Override
+      public void widgetSelected(SelectionEvent e) {
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java]: Nit: `event`.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java [hunk 4]
```diff
@@ -284,15 +286,37 @@ private void createProjectIdSection() {
...
+        openCreateProjectDialog();
+      }
+
+      @Override
+      public void widgetDefaultSelected(SelectionEvent event) {
+        widgetSelected(event);
+      }
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java]: I think you can remove this method if you use `SelectionAdaptor`.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java [hunk 5]
```diff
@@ -284,15 +286,37 @@ private void createProjectIdSection() {
...
+    });
+
     accountSelector.addSelectionListener(new Runnable() {
       @Override
       public void run() {
-        Credential selectedCredential = accountSelector.getSelectedCredential();
-        projectSelector.setProjects(retrieveProjects(selectedCredential));
+        loadProjectsForCredential();
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java]: `refresh` sounds better to me.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java [hunk 6]
```diff
@@ -375,6 +399,16 @@ private Composite createBucketSection(Composite parent) {
     return bucketComposite;
   }
 
+  private void openCreateProjectDialog() {
+    new CreateGcpProjectwithAppengineDialog(this.getShell()).open();
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java]: Probabaly `this` is not needed.

**Comment by chanseokoh** [comment]
There is a test failure that needs to be fixed.

File: plugins/com.google.cloud.tools.eclipse.projectselector/META-INF/MANIFEST.MF (hunk 1 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.projectselector/META-INF/MANIFEST.MF]: It's needed by the Dialog class. Without it we don't get error in Eclipse, but we get a warning that `Window.open()` is restricted and we shouldn't use it. This makes the dependency explicit and Eclipse happy.

Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.projectselector/META-INF/MANIFEST.MF]: This is a `projectselector` bundle, where the only change in the PR is some SWT flags. I think you put `Window.open()` in this bundle but later refactored it to the `.deploy.ui` bundle?

**Comment by briandealwis** [comment]
Here's a screenshot.  The dialog is just a placeholder, right?

screen shot 2017-02-23 at 10 06 20 pm

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java (hunk 5 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java]: Agreed!

File: plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/images/SharedImages.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+import org.eclipse.swt.graphics.Image;
+import org.eclipse.ui.plugin.AbstractUIPlugin;
+
+public class SharedImages {
+
+  private static final ImageDescriptor refreshImageDescriptor =
+      AbstractUIPlugin.imageDescriptorFromPlugin("org.eclipse.search",
+                                                 "icons/full/elcl16/refresh.gif");
+
+  public static Image getRefreshIcon(Device device) {
```
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/images/SharedImages.java]: `createRefreshIcon()`? I suggest to add a Javadoc saying that callers must call `dispose()` after done using the image.

**Comment by akerekes** [comment]
Latest screenshot:

image

**Comment by akerekes** [comment]
@elharo @briandealwis PTAL

**Comment by akerekes** [comment]
@briandealwis 
> I'd find it useful to be able to open a project in the console.

Should we open an issue for this to elaborate more?

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties [hunk 8]
```diff
@@ -40,6 +40,11 @@ tooltip.stop.previous.version=If checked, stops the previously running version w
  new version that receives all traffic.
 tooltip.staging.bucket=The Google Cloud Storage bucket used to stage files for deployment. \
  If not specified, the application''s default code bucket is used.
+
+#Project selector
+projectselector.createproject=Select a GCP Project. Use the Cloud Console to <a href="{0}">create a new project</a>.
+projectselector.createproject.tooltip=Opens the Cloud Console in a browser, you will need to refresh the project list \
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties]: in a browser. You need to refresh

File: plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/images/SharedImages.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+                                                 "icons/full/elcl16/refresh.gif");
+
+  private static final ImageDescriptor refreshPngImageDescriptor =
+      AbstractUIPlugin.imageDescriptorFromPlugin("org.eclipse.search",
+                                                 "icons/full/elcl16/refresh.png");
+
+  /**
+   * Creates an {@link Image} to be used to indicate a 'refresh' action.
+   * <p>
+   * The caller is responsible for properly disposing the Image once it's not needed anymore by
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/images/SharedImages.java]: delete "properly"

**Comment by elharo** [comment]
@hray2003 Please comment on latest UI as shown in above screenshot.

File: plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/images/SharedImages.java [hunk 10]
```diff
@@ -34,7 +34,7 @@
   /**
    * Creates an {@link Image} to be used to indicate a 'refresh' action.
    * <p>
-   * The caller is responsible for properly disposing the Image once it's not needed anymore by
+   * The caller is responsible for disposing the Image once it's not needed anymore by
    * calling {@link Image#dispose()} on the image object.
    * @return the image or null if the image was not found
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/images/SharedImages.java]: blank line before @return

File: plugins/com.google.cloud.tools.eclipse.ui.util/META-INF/MANIFEST.MF [hunk 11]
```diff
@@ -9,7 +9,8 @@ Bundle-Localization: plugin
 Bundle-RequiredExecutionEnvironment: JavaSE-1.7
 Bundle-ActivationPolicy: lazy
 Require-Bundle: org.eclipse.ui.workbench,
- org.eclipse.ui
+ org.eclipse.ui,
+ org.eclipse.search
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.ui.util/META-INF/MANIFEST.MF]: Is the only reason we're pulling in this search package to get the image? If so, I wonder if we might be better off copying it, and then we can remove the logic to choose between GIF and PNG.

File: plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/images/SharedImages.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+  private static final ImageDescriptor refreshPngImageDescriptor =
+      AbstractUIPlugin.imageDescriptorFromPlugin("org.eclipse.search",
+                                                 "icons/full/elcl16/refresh.png");
+
+  /**
+   * Creates an {@link Image} to be used to indicate a 'refresh' action.
+   * <p>
+   * The caller is responsible for disposing the Image once it's not needed anymore by
+   * calling {@link Image#dispose()} on the image object.
+   * @return the image or null if the image was not found
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/images/SharedImages.java]: blank line before @return

File: plugins/com.google.cloud.tools.eclipse.ui.util/META-INF/MANIFEST.MF (hunk 11 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.ui.util/META-INF/MANIFEST.MF]: My thinking here was that the org.eclipse.search bundle will be always there (or not?) so we could just reuse that, it won't increase the installed size of our plugin.

And probably I should file an issue to add the icon the the Eclipse API, I've seen several plugins containing a copy of the icon.

**Comment by hray2003** [comment]
looks great! My 2 cents:

1. Improve UI text for better clarification:
"Select a GCP project." in primary text color.
"If you want to deploy to a new project, create one in Google Cloud Console[link], and refresh the list." Separate line in secondary text color.

2. Align text on the left with same spacing if possible.
screen shot 2017-03-03 at 9 51 44 am

File: plugins/com.google.cloud.tools.eclipse.ui.util/META-INF/MANIFEST.MF (hunk 11 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.ui.util/META-INF/MANIFEST.MF]: Filing an issue sounds like a good idea.

It's a close call, but I prefer copying the image since it also removes the extra if to choose between GIF and PNG; but up to you.

**Comment by elharo** [comment]
This is a nit, but the reload button should be disabled when no account is selected.

File: plugins/com.google.cloud.tools.eclipse.ui.util/META-INF/MANIFEST.MF (hunk 11 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.ui.util/META-INF/MANIFEST.MF]: I kinds agree re: copying the image

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties [hunk 13]
```diff
@@ -42,7 +42,8 @@ tooltip.staging.bucket=The Google Cloud Storage bucket used to stage files for d
  If not specified, the application''s default code bucket is used.
 
 #Project selector
-projectselector.createproject=Select a GCP Project. Use the Cloud Console to <a href="{0}">create a new project</a>.
+projectselector.selectProject=Select a Google Cloud Platform project
+projectselector.createproject=To create a project, open the <a href="{0}">Cloud Console</a> then refresh this list.
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties]: Needs punctuation. I think  a period after the link is probably best:

To create a project, open the <a href="{0}">Cloud Console</a>. Then refresh this list.

Hmm, actually merely opening the console is not enough. How about:

Create a project in the <a href="{0}">Cloud Console</a>. Then refresh this list.

File: plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java [hunk 14]
```diff
@@ -38,12 +38,42 @@ private FontUtil() { }
...
+   * <em>If converting fonts is a frequent operation, this method will create
+   * several {@link DisposeListener}s that can lead to high resource allocation</em>
+   *
+   * @param control whose font will be changed
+   * @param style  e.g. SWT.BOLD or SWT.ITALIC
+   */
+  public static void convertFont(Control control, int style) {
     for (FontData fontData : control.getFont().getFontData()) {
-      if ((fontData.getStyle() & SWT.BOLD) != 0) {
+      if ((fontData.getStyle() & style) != 0) {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java]: This line is hard to follow, and I have personally had really nasty bugs with bitwise constants for fonts like this one before that have cost me days of debugging so it makes me nervous. Could we pull this out top a helper method?

File: plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java [hunk 15]
```diff
@@ -38,12 +38,42 @@ private FontUtil() { }
...
+   * that style.
+   * <p>
+   * If the font is converted, it will attach a {@link DisposeListener}
+   * to the <code>control</code> to dispose the font when it's not needed anymore.
+   * <p>
+   * <em>If converting fonts is a frequent operation, this method will create
+   * several {@link DisposeListener}s that can lead to high resource allocation</em>
+   *
+   * @param control whose font will be changed
+   * @param style  e.g. SWT.BOLD or SWT.ITALIC
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java]: nit: extra space after style

**Comment by akerekes** [comment]
Latest screenshot:

image

File: plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java (hunk 14 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java]: done

File: plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java (hunk 15 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java]: done

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties (hunk 13 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/messages.properties]: done

File: plugins/com.google.cloud.tools.eclipse.ui.util/META-INF/MANIFEST.MF (hunk 11 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.ui.util/META-INF/MANIFEST.MF]: #1532

**Comment by akerekes** [comment]
@elharo PTAL

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java [hunk 16]
```diff
@@ -316,17 +327,50 @@ private void createProjectIdSection() {
...
+    projectSelector = new ProjectSelector(projectSelectorComposite);
+    GridDataFactory.fillDefaults().grab(true, false).hint(SWT.DEFAULT, 200)
+        .applyTo(projectSelector);
+
+    final Button refreshProjectsButton = new Button(projectSelectorComposite, SWT.NONE);
+    refreshProjectsButton.setImage(refreshIcon);
+    GridDataFactory.swtDefaults().align(SWT.END, SWT.BEGINNING).applyTo(refreshProjectsButton);
+    refreshProjectsButton.addSelectionListener(new SelectionAdapter() {
+      @Override
+      public void widgetSelected(SelectionEvent e) {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java]: e --> event

File: plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java [hunk 17]
```diff
@@ -38,12 +38,42 @@ private FontUtil() { }
...
+   */
+  public static void convertFont(Control control, int style) {
     for (FontData fontData : control.getFont().getFontData()) {
-      if ((fontData.getStyle() & SWT.BOLD) != 0) {
+      if (hasStyle(fontData, style)) {
         return;
       }
     }
-    FontDescriptor boldDescriptor = FontDescriptor.createFrom(control.getFont()).setStyle(SWT.BOLD);
+    FontDescriptor boldDescriptor = FontDescriptor.createFrom(control.getFont()).setStyle(style);
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java]: This isn't necessarily bold anymore, right?

File: plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java [hunk 18]
```diff
@@ -54,4 +84,7 @@ public void widgetDisposed(DisposeEvent event) {
     });
   }
 
+  private static boolean hasStyle(FontData fontData, int style) {
+    return (fontData.getStyle() & style) != 0;
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java]: Should this be `== style`?

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java [hunk 19]
```diff
@@ -316,17 +327,50 @@ private void createProjectIdSection() {
...
+        refreshProjectsForSelectedCredential();
+      }
+    });
+
     accountSelector.addSelectionListener(new Runnable() {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java]: This method is really long and complex. It might help to use non-anonymous inner classes instead.

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java [hunk 20]
```diff
@@ -361,6 +367,7 @@ public void widgetSelected(SelectionEvent e) {
       @Override
       public void run() {
         refreshProjectsForSelectedCredential();
+        refreshProjectsButton.setEnabled(accountSelector.getSelectedCredential() != null);
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java]: Does this `refreshProjectButton` need to have its enablement state initially set?

File: plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java (hunk 18 above)
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java]: E.g., if style bits are `0b11` and `getStyle() == 0b10`

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java (hunk 16 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java]: done

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java (hunk 19 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java]: done

File: plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java (hunk 17 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java]: done

File: plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java (hunk 18 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java]: javadoc for `getStyle()`:
>Returns the style of the receiver which is a bitwise OR of one or more of the SWT constants NORMAL, BOLD and ITALIC.

So I think we need the bitwise operation. What do you think?

File: plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java (hunk 20 above)
Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/StandardDeployPreferencesPanel.java]: When databinding is added, the `accountSelector`'s `selectAccount()` is called and that will in turn call this method, so it will initialize the enablement.
However we add databinding for this enablement instead of setting it in the event handler.

File: plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java (hunk 18 above)
Comment (chanseokoh) [inline plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java]: @briandealwis is suggesting `(fondData.getStyle() & style) == style`, which is safe in cases where style bits are not mutually exclusive. But then, it's very unlikely that style bits overlap. However, there's no harm for doing `== style`.

Comment (akerekes) [inline plugins/com.google.cloud.tools.eclipse.ui.util/src/com/google/cloud/tools/eclipse/ui/util/FontUtil.java]: Gotcha! Yeah, they shouldn't overlap. But e.g. if we create a new method that makes the font bold and italic and we pass in those bits in a single int, that could cause a problem similar to @briandealwis 's example.
